flow/layers/image_filter_layer.cc

```diff
@@ -56,37 +66,36 @@ void ImageFilterLayer::Preroll(PrerollContext* context,
   child_bounds = SkRect::Make(filter_output_bounds);
 
   set_paint_bounds(child_bounds);
+}
 
-  transformed_filter_ = nullptr;
+bool ImageFilterLayer::IsNeedCached(PrerollContext* context,
+                                    const SkMatrix& ctm) {
   if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
-    // We have rendered this same ImageFilterLayer object enough
-    // times to consider its properties and children to be stable
-    // from frame to frame so we try to cache the layer itself
-    // for maximum performance.
-    TryToPrepareRasterCache(context, this, matrix);
-  } else {
-    // This ImageFilterLayer is not yet considered stable so we
-    // increment the count to measure how many times it has been
-    // seen from frame to frame.
-    render_count_++;
-
-    // Now we will try to pre-render the children into the cache.
-    // To apply the filter to pre-rendered children, we must first
-    // modify the filter to be aware of the transform under which
-    // the cached bitmap was produced. Some SkImageFilter
-    // instances can do this operation on some transforms and some
-    // (filters or transforms) cannot. We can only cache the children
-    // and apply the filter on the fly if this operation succeeds.
-    transformed_filter_ = filter_->makeWithLocalMatrix(matrix);
-    if (transformed_filter_) {
-      // With a modified SkImageFilter we can now try to cache the
-      // children to avoid their rendering costs if they remain
-      // stable between frames and also avoiding a rendering surface
-      // switch during the Paint phase even if they are not stable.
-      // This benefit is seen most during animations.
-      TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
-    }
+    return true;
+  }
+  transformed_filter_ = nullptr;
+  // This ImageFilterLayer is not yet considered stable so we
+  // increment the count to measure how many times it has been
+  // seen from frame to frame.
+  render_count_++;
+
+  // Now we will try to pre-render the children into the cache.
+  // To apply the filter to pre-rendered children, we must first
+  // modify the filter to be aware of the transform under which
+  // the cached bitmap was produced. Some SkImageFilter
+  // instances can do this operation on some transforms and some
+  // (filters or transforms) cannot. We can only cache the children
+  // and apply the filter on the fly if this operation succeeds.
+  transformed_filter_ = filter_->makeWithLocalMatrix(ctm);
+  if (transformed_filter_) {
+    // With a modified SkImageFilter we can now try to cache the
+    // children to avoid their rendering costs if they remain
+    // stable between frames and also avoiding a rendering surface
+    // switch during the Paint phase even if they are not stable.
+    // This benefit is seen most during animations.
+    return true;
```

Comment:
The TryToPrepare call here was preparing the CacheableChild(), not *this*!

So, perhaps the layer-style of cache record needs to have a flag for "cache me" or "cache my children". In either case, we'd skip the same amount of children.

Reply:
It could replace the layer in the record with GetCacheableChild in this case.

Another option is to have a special CacheableChild record that has a flag for "me" or "my cacheable child". Either way, there is a PR outstanding that will try to cache the children of a layer without a CacheableChild, so this gets a little complicated.

The best choice for now would be to just replace the layer pointer with the cacheable child in this case.